packages/material-ui/src/utils/unstable_useId.js

```diff
@@ -1,18 +1,22 @@
 import * as React from 'react';
 
+/**
+ * @ignore - internal component.
+ */
+export const IdContext = React.createContext({
+  generateId: () => `mui-${Math.round(Math.random() * 1e5)}`,
+});
+
+if (process.env.NODE_ENV !== 'production') {
+  IdContext.displayName = 'IdContext';
+}
+
 /**
  * Private module reserved for @material-ui/x packages.
  */
 export default function useId(idOverride) {
-  const [defaultId, setDefaultId] = React.useState(idOverride);
+  const idContext = React.useContext(IdContext);
+  const [defaultId] = React.useState(idOverride || idContext.generateId());
```

Comment:
Wouldn't this cause issues in SSR and/or Strict Mode / CM where without `IdContext.Provider` this non-deterministically generates a random ID on first render? In other words, this implementation seems to go against what the deleted comment below was warning about?

Reply:
Right, there was something that was off about context for ids and I couldn't figure out what it was. We would need to restrict `idContext` to be side-effect free. And from my experience this isn't something that is doable for these kind of functions. @nickpalmer Could you include an example implementation of a custom `generateId`?

Reply:
I think we need an SSR test to accompany this change. Or maybe if I get time this weekend, I'll raise an SSR test for the existing implementation.

Reply:
I don't see how the existing implementation doesn't cause problems in SSR since the IDs it generates at random which is why it breaks snapshot testing. The comment was implying that the function cannot be used in SSR, which is ALSO a problem. At least this code doesn't break it.
```
let id = 0;
const generateId = () => `id-${id}`;
return React.renderToString(<IdContext.Provider value={{ generateId }}><App /></IdContext.Provider>)
```
I will try to fixup this PR and the docs tomorrow, but work work has been a bit busy and I am still learning the documentation system.

Reply:
> I don't see how the existing implementation doesn't cause problems in SSR since the IDs it generates at random

It happens inside an effect and effects do not run on the server.